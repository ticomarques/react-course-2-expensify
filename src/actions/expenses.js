/*
Aqui ficam as acoes relativas aos expenses, que sao enviadas ao reducer junto com o state

Em casa acao ficam objetos planos, dos valores que serão enviados via parametro usando as funcoes, tipo:

No componente tem o dispatch :

dispatch(removeExpense({ id: props.expense.id }));

A funcao removeExpense, mora aqui no arquivo de actions, ele contém :

1- a funcao removeExpense()
2- o parametro passado para funcao ({id} = {}), onde {id} = {} significa se não enviar valor 
para o parametro id o valor default é {} objeto vazio
3- dentro da funcao tempos o type, que vai servir para linkar a funcao action com o reducer
4- o valor que vai ser usado para adicao, edicao ou exclusao.

Actions é tipo local que o dispatch manda a informacao primeiro via paramtro, depois vai para o reducer junto com state

*/

import uuid from 'uuid';

// ACTION - ADD_EXPENSE
export const addExpense = (
  {
    description = '',
    note = '',
    amount = 0,
    createdAt = 0
  } = {}
) => ({
  type: 'ADD_EXPENSE',
  expense: {
    id: uuid(),
    description,
    note,
    amount,
    createdAt
  }
});

// ACTION - REMOVE_EXPENSE
export const removeExpense = ({ id } = {}) => ({
  type: 'REMOVE_EXPENSE',
  id
});

// ACTION - EDIT_EXPENSE
export const editExpense = (id, updates) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates
});
