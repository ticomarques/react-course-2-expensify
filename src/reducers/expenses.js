// Expenses Reducer

const expensesReducerDefaultState = [];

export default (state = expensesReducerDefaultState, action) => {
  switch (action.type) {
    case 'ADD_EXPENSE':
      return [
        ...state,
        action.expense
      ];//usamos array destructuring para adicionar novo item ao array de expenses QUE ESTA NO STATE
    case 'REMOVE_EXPENSE':
      return state.filter(({ id }) => id !== action.id);
      //Usamos o filter para criar um novo array com todos os itens com ID diferentes do passado
    case 'EDIT_EXPENSE':
      return state.map((expense) => {
        if (expense.id === action.id) {
          return {
            ...expense,
            ...action.updates
          };
        } else {
          return expense;
        };
        //Usamos o map para varrer o array e achar o indice com id igual ao action id, 
        //e depois atualiza-lo com action.updates
      });
    case 'SET_EXPENSES':
      return action.expenses;
    default:
      return state;
  }
};
