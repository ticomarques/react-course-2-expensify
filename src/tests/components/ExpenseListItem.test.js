import React from 'react';
import { shallow } from 'enzyme';
import expenses from '../fixtures/expenses';
import ExpenseListItem from '../../components/ExpenseListItem';

test('Render ExpenseListItem correctly', () => {
    const wrapper = shallow(<ExpenseListItem {...expenses[0]} />);
    expect(wrapper).toMatchSnapshot();
});
