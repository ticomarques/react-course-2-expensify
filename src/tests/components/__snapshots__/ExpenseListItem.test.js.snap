// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render ExpenseListItem correctly 1`] = `
<div>
  <h3>
    <Link
      replace={false}
      to="/edit/1"
    >
      Gum
    </Link>
  </h3>
  <p>
    $1.95
    -
    December 31st, 1969
  </p>
</div>
`;
